### PR TITLE
Make DSACryptoServiceProvider set legal key sizes to most restrictive for platform

### PR DESCRIPTION
**Pull request by elinor-fung**
The legal key sizes for DSA are more restrictive than on other platforms. Update `DSACryptoServiceProvider` to use the most restrictive sizes as the legal key sizes.

cc @jkoritzinsky @steveisok @AaronRobinsonMSFT @bartonjs

### CODE REVIEW

**Review by jkoritzinsky — APPROVED** [review]

**Review by AaronRobinsonMSFT — APPROVED** [review]

File: src/libraries/Common/tests/System/Security/Cryptography/AlgorithmImplementations/DSA/DSASignVerify.cs [hunk 1]
```diff
@@ -163,7 +163,17 @@ private void UseAfterDispose(bool importKey)
             Assert.Throws<ObjectDisposedException>(
                 () =>
                 {
-                    key.KeySize = 576;
+                    try
+                    {
+                        key.KeySize = 576;
+                    }
+                    catch (CryptographicException) when (PlatformDetection.IsAndroid)
```
Comment (steveisok) [inline src/libraries/Common/tests/System/Security/Cryptography/AlgorithmImplementations/DSA/DSASignVerify.cs:170]: Exception filters are problematic in AOT scenarios.  I would move this check inside the catch block.

Comment (bartonjs) [reply src/libraries/Common/tests/System/Security/Cryptography/AlgorithmImplementations/DSA/DSASignVerify.cs:170]: Do we AOT our tests?

File: src/libraries/System.Security.Cryptography.Csp/src/System/Security/Cryptography/DSACryptoServiceProvider.Unix.cs [hunk 2]
```diff
@@ -12,27 +12,41 @@ namespace System.Security.Cryptography
...
 
+            // Depending on the platform, _impl's legal key sizes may be more restrictive than Windows provider.
+            //   DSAAndroid : (1024, 3072, 1024)
+            //   DSAOpenSsl : (512, 3072, 64)
+            //   DSASecurityTransforms : (512, 1024, 64)
+            //   Windows CSP : (512, 1024, 64)
+            // Use the most restrictive legal key sizes
+            LegalKeySizesValue = OperatingSystem.IsAndroid()
+                ? s_legalKeySizesAndroid
+                : s_legalKeySizesWindowsCsp;
```
Comment (bartonjs) [inline src/libraries/System.Security.Cryptography.Csp/src/System/Security/Cryptography/DSACryptoServiceProvider.Unix.cs:48]: Save this computation as a static readonly so it only has to evaluate once?  (Sure, this type isn't popular, but may as well stay in good habits).

File: src/libraries/System.Security.Cryptography.Csp/src/System/Security/Cryptography/DSACryptoServiceProvider.Unix.cs [hunk 3]
```diff
@@ -147,14 +161,14 @@ public override int KeySize
             set
             {
                 // Perform the check here because LegalKeySizes are more restrictive here than _impl
-                if (!value.IsLegalSize(s_legalKeySizes))
+                if (!value.IsLegalSize(this.LegalKeySizesValue!))
```
Comment (bartonjs) [inline src/libraries/System.Security.Cryptography.Csp/src/System/Security/Cryptography/DSACryptoServiceProvider.Unix.cs:164]: The `this.` shouldn't be required.  Also, if you save the result of the OS-based ternary to `s_legalKeySizes`, this line can be reverted.

File: src/libraries/System.Security.Cryptography.Csp/src/System/Security/Cryptography/DSACryptoServiceProvider.Unix.cs [hunk 4]
```diff
@@ -147,14 +161,14 @@ public override int KeySize
...
                     throw new CryptographicException(SR.Cryptography_InvalidKeySize);
 
                 _impl.KeySize = value;
             }
         }
 
-        public override KeySizes[] LegalKeySizes => s_legalKeySizes.CloneKeySizesArray();
+        public override KeySizes[] LegalKeySizes => base.LegalKeySizes;
```
Comment (bartonjs) [inline src/libraries/System.Security.Cryptography.Csp/src/System/Security/Cryptography/DSACryptoServiceProvider.Unix.cs:171]: This line could also be reverted if the ternary is saved to `s_legalKeySizes`

File: src/libraries/Common/tests/System/Security/Cryptography/AlgorithmImplementations/DSA/DSASignVerify.cs (hunk 1 above)
Comment (steveisok) [reply src/libraries/Common/tests/System/Security/Cryptography/AlgorithmImplementations/DSA/DSASignVerify.cs]: Yes, we send all assemblies in the publish directory, for example, to the AOT compiler.

Comment (bartonjs) [reply src/libraries/Common/tests/System/Security/Cryptography/AlgorithmImplementations/DSA/DSASignVerify.cs]: Huh, TIL.  Cool :smile:.

**Review by bartonjs — APPROVED** [review]

**Pull request merged**